signature-v4: Use sha256("") for calculating canonical request

krishnasrinivas opened the pull request:
## Description
<!--- Describe your changes in detail -->

When x-amz-content-sha256 header is missing assume e3b0c44298fc1c149afbf4c8996fb92427ae41e4649b934ca495991b7852b855
as the SHA256 value for calculating canonical-request.

<!--- Provide a general summary of your changes in the Title above -->


## Motivation and Context
<!--- Why is this change required? What problem does it solve? -->
<!--- If it fixes an open issue, please link to the issue here. -->
fixes #4039

## How Has This Been Tested?
<!--- Please describe in detail how you tested your changes. -->
<!--- Include details of your testing environment, and the tests you ran to -->
<!--- see how your change affects other areas of the code, etc. -->
make test

## Types of changes
<!--- What types of changes does your code introduce? Put an `x` in all the boxes that apply: -->
- [x] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist:
<!--- Go over all the following points, and put an `x` in all the boxes that apply. -->
<!--- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->
- [ ] My code follows the code style of this project.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [x] I have added tests to cover my changes.
- [x] All new and existing tests passed.

krishnasrinivas committed:
signature-v4: Use sha256("") for calculating canonical request

fixes #4039

When x-amz-content-sha256 header is missing assume e3b0c44298fc1c149afbf4c8996fb92427ae41e4649b934ca495991b7852b855
as the SHA256 value for calculating canonical-request.

harshavardhana commented on cmd/signature-v4-utils.go:
```diff
@@ -47,6 +47,22 @@ func skipContentSha256Cksum(r *http.Request) bool {
 
+// Returns SHA256 for calculating canonical-request.
+func getContentSha256Cksum(r *http.Request) string {
+	defaultSum := emptySHA256
```
This function is pretty confusing.. needs work. the relation between defaultSum and sum is not apparent.

krishnasrinivas replied:
do you have suggestions?

harshavardhana replied:
```go
// Returns SHA256 for calculating canonical-request.
func getContentSha256Cksum(r *http.Request) string {
        // For a presigned request we look at the query param for sha256.
	if isRequestPresignedSignatureV4(r) { 
                 presignedCkSum := r.URL.Query().Get("X-Amz-Content-Sha256")
                 if presignedCkSum == "" {
                        // If not set presigned is defaulted to UNSIGNED-PAYLOAD.
                        return unsignedPayload
                 }
                 return presignedCkSum
         }
	contentCkSum := r.Header.Get("X-Amz-Content-Sha256")
        if contentCkSum == "" {
                // If not set content checksum is defaulted to sha256([]byte("")). 
                contentCkSum = emptySHA256
        }
	return contentCkSum
}
```

krishnasrinivas committed:
Make getContentSha256Cksum more readable.
Use X-Amz-Content-Sha256 for SHA verification instead of getContentSha256Cksum() return value.

krisis commented on cmd/auth-handler.go:
```diff
@@ -142,52 +142,55 @@ func isReqAuthenticatedV2(r *http.Request) (s3Error APIErrorCode) {
+		sum = r.URL.Query().Get("X-Amz-Content-Sha256")
 	}
-	return ErrAccessDenied
+	if sum != "" && sum != getSHA256Hash(payload) {
```
when can `sum` be `""`? Could you add a comment for it?

donatello replied:
Looks like the above code can just use `getContentSha256Cksum` - it need not be reimplemented here.

krishnasrinivas replied:
`getContentSha256Cksum()` returns `e3b0c44298fc1c149afbf4c8996fb92427ae41e4649b934ca495991b7852b855` when `x-amz-content-sha256` is missing and hence should not be used, else we'll end up returning "sha256 does not match" error.

krishnasrinivas replied:
> when can sum be ""? Could you add a comment for it?

@krisis is comment needed because, then I will have to add a comment "When X-Amz-Content-Sha256 header is missing" but the code for it seems obvious:

```
+	sum := r.Header.Get("X-Amz-Content-Sha256")
 +	if isRequestPresignedSignatureV4(r) {
 +		sum = r.URL.Query().Get("X-Amz-Content-Sha256")
  	}
```

krisis replied:
>// Verify that X-Amz-Content-Sha256 Header == sha256(payload)

@krishnasrinivas the comment above is obvious too. But it informs the reader that sha256sum of payload needs to be validated. In the same spirit you can add to the above comment, what is to be done when sha256sum of payload is not sent as part of the request.

krishnasrinivas replied:
@krisis do you mean like this:

```
// Verify that X-Amz-Content-Sha256 Header == sha256(payload)
// If X-Amz-Content-Sha256 header is not sent then we don't calculate/verify sha256(payload)
```

If it is not helping then you can suggest a comment.

krisis replied:
This works.

krisis commented on cmd/signature-v4_test.go:
```diff
@@ -135,21 +135,7 @@ func TestDoesPresignedSignatureMatch(t *testing.T) {
 			region:   "us-west-1",
 			expected: ErrInvalidAccessKeyID,
 		},
-		// (2) Should error when the payload sha256 doesn't match.
```
why is this test case being removed?

krishnasrinivas replied:
sha256 verification is not done in this function, just signature verification happens here.

donatello commented on cmd/signature-v4-utils.go:
```diff
@@ -47,6 +47,27 @@ func skipContentSha256Cksum(r *http.Request) bool {
+		if presignedCkSum == "" {
+			// If not set presigned is defaulted to UNSIGNED-PAYLOAD.
+			return unsignedPayload
+
```
Stray newline.

donatello commented on cmd/signature-v4-utils.go:
```diff
@@ -47,6 +47,27 @@ func skipContentSha256Cksum(r *http.Request) bool {
+	if contentCkSum == "" {
+		// If not set content checksum is defaulted to sha256([]byte("")).
+		contentCkSum = emptySHA256
+
```
Stray newline.

donatello commented on cmd/auth-handler_test.go:
```diff
@@ -308,6 +308,16 @@ func mustNewSignedRequest(method string, urlStr string, contentLength int64, bod
+	req.Header.Set("Content-Md5", "YWFhYWFhYWFhYWFhYWFhCg==")
+	cred := serverConfig.GetCredential()
+	if err := signRequestV4(req, cred.AccessKey, cred.SecretKey); err != nil {
+		t.Fatalf("Unable to inititalized new signed http request %s", err)
```
Typo: `s/inititalized/initialized/`

krishnasrinivas committed:
Review fixes.
